### PR TITLE
[WIP] Fix malformed output dictionary in HashProbe::evalFilter()

### PR DESCRIPTION
**Pull request by karteekmurthys**
`HashProbe::evalFilter()` while rewriting the output dictionary vector copies the previous batch `noMatchDetector_.currentRow` as the first entry into the output dictionary. This increases the size of the dictionary to be one more than previously reserved size leading to this error:

```
C++ exception with description "Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: (40 vs. 44) Malformed dictionary, index array is shorter than DictionaryVector
Retriable: False
Expression: dictionaryIndices->size() >= length * sizeof(vector_size_t)
Function: DictionaryVector
File: /Users/karteek/workspace/kart-presto/presto-native-execution/velox/velox/vector/DictionaryVector-inl.h
Line: 92
```
Added a test to reproduce the issue.

### CODE REVIEW

File: velox/exec/tests/HashJoinTest.cpp [hunk 1]
```diff
@@ -5751,6 +5751,49 @@ DEBUG_ONLY_TEST_F(HashJoinTest, hashBuildAbortDuringInputgProcessing) {
   }
 }
 
+TEST_F(HashJoinTest, malformedDictEvalFilter) {
+  /*
+   * The last row of previous batch did pass the filter.
```
Comment (Krishna-Prasad-P-V) [inline velox/exec/tests/HashJoinTest.cpp:5756]: The last row of the previous batch did `not` pass the filter (false).

This should be the case if it needs to be the first row of the next batch.

**Comment by aditi-pandit** [comment]
@kevinwilfong : Wanted to brainstorm a bit on the solution here.
So the problem is as follows:
In getOutput() each batch of HashProbe input rows are handled at a time. 

- For each batch, the matching rows are found from the HashTable. 
- Then evalFilter(...) is invoked to apply the filter.
- For left outer joins, rows that do not pass the filter are tracked and output with null values on rhs.

If the last row of a batch of input corresponds to a value for which the filter has always returned false then a decision cannot be taken about this output row within the current getOutput. The next batch of rows is inspected. If the first row matches the last row then the logic is alright. But if the new row doesn't match the old one, then the old one needs to be output.

Because the logic in evalFilter works by overwriting rawOutputProbeMapping and outputTableRows_, the old value overwrites new ones and the new ones are not available for further processing. The overwritten old value is copied to all the rows instead. Since we have also lost the input vectors, I'm not sure how to reconstruct the input row. The row value we have could be corrupt. But anyways, the malformed dictionary failure is encountered since there is one additional unexpected row.

So there are multiple parts for the solution:
i) Retain the last input row. We would have to decode the contents of this row into a FlatVector and retain it across getOutput() calls at the very least.
ii) Detect that such a row is pending. This is fairly straightforward with the fields in noMatchDetector_.
iii) This output row would need to be set aside from the current input buffer processing. The row from the previous batch interferes with how the current batch is handled. So its better to let the current batch be handled as is.
iv) Where do we return this output row then ? If the current output block is handled as a dictionary, then its hard to accommodate this row in it. It might be better to keep such "last rows" in a batch of a FlatVector and output it at the end.

What do you think ? Do you have any other ideas ?

The unit test in this PR faithfully reproduces the problem we have encountered.

**Comment by mbasmanova** [comment]
@aditi-pandit Somehow, malformedDictEvalFilter test is passing for me. I also don't see it failing in CI. Is this the right repro?

**Comment by mbasmanova** [comment]
@aditi-pandit Didn't realize this PR includes a fix as well. Reverted the fix and I can reproduce the issue now.

**Comment by mbasmanova** [comment]
@aditi-pandit Aditi, I took a look and would like to clarify a few things.

HashProbe processing probe input one batch at a time.

For each batch of probe input, getOutput() is called multiple times. 

Let's say probe input batch has M rows and when joined with the build side the total number of rows is M'. We process M' matches N rows at a time. (N = outputBatchSize_).

For each N matches, we evaluate the filter to identify "true" matches. We then produce output for all probe input rows exception the "last" (as you pointed out, in case of a miss we need to check next batch of N matches before producing output for that probe row). Hence, it is possible that subsequent output batch will have N + 1 row. I believe the fix here accommodates this scenario.

I do not see any further issues. Do you?

**Comment by karteekmurthys** [comment]
@mbasmanova I was just trying out the fix in this issue by allocating 1 more than outputBatchSize. This somehow fixes the symptoms but not the underlying issue. I have explained the root cause here: https://github.com/facebookincubator/velox/issues/6063#issuecomment-1720775964
1. All rows in `rawOutputRowMapping` is overwritten with the same row number saved from previous batch.
2. Like you mentioned the outputbatchsize is not enough to hold the final result.

The fix is more involved than just increasing the outputTableRows/rawOutputRowMapping batch size.

**Comment by kevinwilfong** [comment]
@aditi-pandit @karteekmurthys 
I've been looking at this as well.  The alternate solution I came up with is to carry the last row forward to the next batch if we find ourselves wanting to create a batch of size N + 1 (similar to the way we carry a miss forward, but this works for a hit as well).

We can store the outputTableRow char* and rowMapping vector_size_t for the final row in the batch and append it to a future batch with available space, or create an additional output batch with just that row if we run out of input.

The input row looks like it's still valid since addInput won't be called again until we finish processing the current batch.  There's also not much issue adding it to the dictionary, again since the input is still the same.

I also had to create some additional buffers for outputTableRows_ and outputRowMapping_ since they're updated in place, but I can probably avoid that.

I think it's easier than peeking ahead, but it's still unfortunately complicated.  In particular, whether or not we're done with an input batch is no longer as simple as checking results_.atEnd(), we now have to remember to check if we're carrying over a row.  This seems unintuitive and error prone.

Personally, I prefer the solution of just allocating space for 1 additional row in the output batch.  It seems safer and simpler, and I don't see anywhere where we guarantee the size of an output batch will have at most outputBatchSize_ rows (it can certainly have less).  outputBatchSize_ comes from either preferredOutputBatchRows or preferredOutputBatchBytes in the QueryConfig, neither of which sound like strong requirements.

**Comment by aditi-pandit** [comment]
@kevinwilfong, @mbasmanova :
Carrying the row forward is reasonable. I feel allocating the additional row in the output batch would be alright. 

I'm unsure if it was sufficient to carry forward outputTableRow char* and rowMapping vector_size_t to reconstruct the final row. I thought we might have to decode the input row contents. But like you are suggesting its all the same dictionary, so the proposal might work.

Did you prototype a solution ? We can try this approach in any case.

**Comment by kevinwilfong** [comment]
I published a draft of the prototype here https://github.com/facebookincubator/velox/pull/6680

It passes all tests under velox_exec_test but I haven't tested it otherwise.

**Comment by kevinwilfong** [comment]
In case it wasn't clear, when I said "I prefer the solution of just allocating space for 1 additional row in the output batch" I was referring to Karteek's previous solution here.  If it works, it's much simpler and cleaner than mine.

**Comment by karteekmurthys** [comment]
> In case it wasn't clear, when I said "I prefer the solution of just allocating space for 1 additional row in the output batch" I was referring to Karteek's previous solution here. If it works, it's much simpler and cleaner than mine.

The solution I posted here is not complete. It was just a hack (so reverted), we still have the issue of overwriting existing rawOutputMapping_ with numDetector_.currentRow which causes this issue: https://github.com/facebookincubator/velox/issues/6063#issuecomment-1720775964

1. We must carry forward the `noMatchDetector_.currentRow` until we find space in a batch to squeeze it in (similar to your draft soluntion here) OR
2. Re-include `noMatchDetector_.currentRow` in the next batch as the first row of input to `evalFilter()`.

@kevinwilfong

**Comment by kevinwilfong** [comment]
Ah, ok, I got it, it's not as simple as just increasing the row mapping size by 1.  I was hitting the same issue with overwriting rawOutputMapping_ when I still needed the data.

Let me clean up my change, and I'll publish it.

**Pull request closed without merging**